CIPs/CIP-79/cip-79.md

```diff
@@ -0,0 +1,270 @@
+    "qYfZN2QNDgjJpaL": "z6LSnTyV1nSWfMmfV1PoRmo3enhDnfqfFqYfZN2QNDgjJpaL"
+  }
+}
+```
+
+To convert this into a DID document first create an empty DID document:
+
+```json
+{
+  "@context": "https://w3id.org/did/v1",
+  "id": did,
+  "publicKey": [],
+  "authentication": [],
+  "keyAgreement": []
+  }
+```
+
+Now iterate though the entires in the `publicKeys` object in the Ceramic document and do the following:
+
+* If it's a `secp256k1` key:
+
+  * Add a `Secp256k1VerificationKey2018` to the *publicKey* array:
+
+    ```js
+    {
+      id: "<did>#<entry-key>",
+      type: "Secp256k1VerificationKey2018",
+      controller: "<did>",
+      publicKeyBase58: "<entry-value-public-key-base58btc-encoded>"
+    }
+    ```
+
+  * Add a `Secp256k1SignatureAuthentication2018` to the *authentication* array:
+
+    ```js
+    {
+      id: "<did>#<entry-key>",
+      type: "Secp256k1SignatureAuthentication2018",
+      controller: "<did>",
+      publicKeyBase58: "<entry-value-public-key-base58btc-encoded>"
+    }
+    ```
+
+* If it's a `x25519` key:
+
+  * Add a `Curve25519EncryptionPublicKey` to the *publicKey* array:
+
+    ```js
+    {
+      id: "<did>#<entry-key>",
+      type: "Curve25519EncryptionPublicKey",
+      controller: "<did>",
+      publicKeyBase58: "<entry-value-public-key-base58btc-encoded>"
+    }
+    ```
+
+  * Add a `X25519KeyAgreementKey2019` to the *keyAgreement* array:
+
+    ```js
+    {
+      id: "<did>#<entry-key>",
+      type: "Secp256k1SignatureAuthentication2018",
+      controller: "<did>",
+      publicKeyBase58: "<entry-value-public-key-base58btc-encoded>"
+    }
+    ```
+
+##### 3IDv0 genesis
+
+Since the content of the Ceramic document for a 3IDv0 is empty we create the DID document by taking the public keys in the 3IDv0 genesis object (see [Appendix A](#appendix-a)) and convert them to multicodec public keys (one `secp256k1` and one `x25519`). The key properties in the DID document should be constructed as above, with the *entry-key* being the last *15* characters of the multicodec encoded keys.
+
+##### DID Document Metadata
+
+When resolving the DID document [DID Document Metadata](https://w3c.github.io/did-core/#did-document-metadata) should be provided. When resolving a 3ID we should populate the following fields:
+
+* `create` - should be populated using the blockchain timestamp of the first *AnchorCommit*
+* `updated` - should be populated using the blockchain timestamp of the most recent *AnchorCommit*
+* `versionId` - should be equal to the commit CID of the most recent *AnchorCommit*
+
+#### Resolving using the `versionId` parameter
+
+When the `versionId` query parameter is given as a DID is resolved it means that we should try to resolve a specific version of the DID document. The resolution process is the same except that the *AnchorCommit* we use for the state of the document should be equal to the DocID + CID from `versionId`. In addition we should construct the *DID Document Metadata* differently.
```

Comment:
We also need to take the **content** from that version of the document, not just the `AnchorCommit` metadata

Reply:
Not sure who exactly the intended audience here is.  For us it's probably obvious that we want to actually load the document contents at that specific version, but if this is for someone from w3c that may not be as obvious.  If someone who wasn't us tried to implement this off of the spec, I could see them loading the current version of the document, but then just iterating the log and taking the `AnchorCommit` metadata as you specify here, but leaving the `content` from the current version.  So that's why I'm being extra pedantic and pointing this out.  If you think it's not a concern for how this doc will be used, then that's fine and feel free to leave it as-is!

Reply:
> The resolution process is the same except that the *AnchorCommit* we use for the **state** of the document should be equal to the DocID + CID from `versionId`.

Imo, **state** implies the *content*. 

The audience of this is really someone that will be implementing the 3ID did method.

Reply:
Clarified this to be more clear about *content*.

Reply:
Okay, if you're not worried about it then I won't either